### ISSUE TITLE
Phase11-Part1-Mohammad-Kasaei

### ISSUE THREAD

**Issue by MKasaei00** [Star-Academy/codestar-intern-issues#189]
This issue template is based of [Phase11-Web.md](https://github.com/Star-Academy/codestar-internship/blob/master/Projects/Phase11-Web/Phase11-Web.md) contents.
Click on "Raw" button on right side of this page to see original markdown then create a new issue with original content of this tasklist:

# Phase 11 First Part

- [x] Before starting: Read the link about history and purpose of creating & developing Web, then answer this questions:
  - [x] How did developers provide server-client softwares for users before creation of web technologies?
  
      `For any user interaction a request was sent to server and rendered in the client device(because the client device was not sophisticated the content was very simple)`
      
  - [x] What was the main trouble with traditional technologies before web?
  
      `Any interaction and change in the page came from server and it was slow`
      
  
  - [x] How did web overcome this main trouble?
  
      `With client side technology advances data the client software could download the data all in one step and show the proper format to the user when required`

## Asp.net core

- [x] Getting Started with HTTP:
  - [x] What does a protocol mean?
  
    `A set of rules that help software systems to interact with each other and understand messages`

  - [x] Why designing a protocol for communication is important prerequirement of all other web related technologies?
  
    `Since web is actually a set of connection between different hosts(devices) it is all about communication and without protocol agreements hosts can not interact with each other`
    
  - [x] Try to get familier with HTTP by reading links.
  - [x] Why HTTP is most popular protocol in web technologies?
    
    `Because HTTP is used over WWW and almost any content on the web is served using this protocol and have different capabilities that can be used in various use cases.`
    
  - [x] What is the most interesting aspect of HTTP in your opinion? 
  
    `I think combining different HTTP verbs with custom headers can create a magic in architecture design`

- [x] Getting Started with REST:
  - [x] Read mentioned links.
  - [x] Why do we need RESTful APIs?
  
    - `Every website have a front-end client side application that will use the backend server to have access to data`
    - `Some websites also have mobile applications so its necessary to serve a connection between backend and app`
    - `We need a popular and generic way of data transfer to present data widely on internet`

- [x] Getting Started with Asp.net Core:
  - [x] Read mentioned links.
  - [x] Create a hello world API.
  - [x] Call your api from browser.
  - [x] Play with routing options! (Try to use it & enjoy it!)
  - [x] Why do we need configurable routings in web development?
  
    `In out backend application different resources exist and we have to server different actions(CRUD) so we must use route to address the resources and their relationship can be shown in nested addresses.`
  
  - [x] Play with dependency injection! (Define an interface & inject interface implementation to code)
  - [x] Why do we need dependency injection?
     - To separate the implementation and interface ( we are working with the interface )
     - To easily manage resources in the application and run efficiently
     - To easily and almost automatically manager object lifecycle and easily retrieve them
  - [x] Play with parameter binding! (use different modes)
  - [x] Install postman.
  - [x] Call your APIs with postman.
  
- [x] Implement Asp.net Core Project:
  - [x] Design your search API.
  - [x] Implement & Test your API.

# Review
Link of your own PR:

[My PR](https://github.com/Star-Academy/Winter99-Codestar-Mohammad-Kasaei/pull/12)

Link of other PR(s) you agreed to review:

`Other PRs`

- [x] Your own PR is reviewed and approved by at least one other team.

- [x] Your own PR is reviewed and approved by mentor (hamyar).

- [x] Your own PR is reviewed and approved by manager (rahbar).

- [x] Your own PR is merged.
